lms/models/grouping.py

```diff
@@ -0,0 +1,71 @@
+import sqlalchemy as sa
+from sqlalchemy.dialects.postgresql import JSONB
+
+from lms.db import BASE
+from lms.models import TimestampedModelMixin
+from lms.models.application_settings import ApplicationSettings
+
+MAX_GROUP_NAME_LENGTH = 25
+
+
+class Grouping(TimestampedModelMixin, BASE):
+    __tablename__ = "grouping"
+    __mapper_args__ = {"polymorphic_identity": "grouping", "polymorphic_on": "type"}
+    __table_args__ = (
+        sa.UniqueConstraint("application_instance_id", "authority_provided_id"),
+    )
+
+    id = sa.Column(sa.Integer(), autoincrement=True, primary_key=True)
+
+    application_instance_id = sa.Column(
+        sa.Integer(),
+        sa.ForeignKey("application_instances.id", ondelete="cascade"),
+        nullable=False,
+    )
+    application_instance = sa.orm.relationship("ApplicationInstance")
+
+    authority_provided_id = sa.Column(sa.UnicodeText(), nullable=False)
+
+    parent_id = sa.Column(
+        sa.Integer(),
+        sa.ForeignKey("grouping.id", ondelete="cascade"),
+        nullable=True,
+    )
+
+    lms_id = sa.Column(sa.String(), nullable=True)
```

Comment:
This can be nullable to allow migrating existing courses to the new table

Reply:
I'm not sure about this column. There are multiple IDs involved here:

* The `context_id` launch param is a unique identifier for the course. This is a standard LTI launch param, I _think_ it works the same across all LMS's (they all use it as a unique course ID)

* In the Canvas API courses also have an `id`. This is not the same as the `context_id`! It's a different ID. Our https://lms.hypothes.is/config_xml uses `$Canvas.course.id` so that we receive the Canvas API course ID as a `custom_canvas_course_id` launch parameter. We need this `custom_canvas_course_id` to make API requests about the course, you can't use the `context_id` in API requests

* Canvas sections have their own ID that's only unique within the course

* I'm guessing Canvas groups also have their own ID that's only unique within the course

I've got a feeling we might want to store all for IDs separately

Reply:
There are some other course-related launch params: `context_type`, `context_label` and `context_title`. I wonder if we want to store those as well? Although apparently we decided not to when we created the original `course` table